ESPixelStick/src/output/OutputUart.hpp

```diff
@@ -171,49 +151,66 @@ inline void IRAM_ATTR StartNewDataFrame()
     uint32_t RxIsr = 0;
     uint32_t ErrorIsr = 0;
     uint32_t IsrIsNotForUs = 0;
-    uint32_t IntensityBytesSent = 0;
+    uint32_t IntensityValuesSent = 0;
     uint32_t IntensityBitsSent = 0;
-    uint32_t IntensityBytesSentLastFrame = 0;
+    uint32_t IntensityValuesSentLastFrame = 0;
     uint32_t IntensityBitsSentLastFrame = 0;
     uint32_t IncompleteFrame = 0;
     uint32_t IncompleteFrameLastFrame = 0;
     uint32_t EnqueueCounter = 0;
+    uint32_t BreakIsrCounter = 0;
+    uint32_t IdleIsrCounter = 0;
+    uint32_t TimerIsrCounter = 0;
+    uint32_t TimerIsrNoDataToSend = 0;
+    uint32_t TimerIsrSendData = 0;
+    uint32_t FiFoNotEmpty = 0;
+    uint32_t FiFoEmpty = 0;
+
 #endif // def USE_UART_DEBUG_COUNTERS
 
-#define DisableUartInterrupts CLEAR_PERI_REG_MASK(UART_INT_ENA(OutputUartConfig.UartId), UART_TXFIFO_EMPTY_INT_ENA);
-#define EnableUartInterrupts  SET_PERI_REG_MASK  (UART_INT_ENA(OutputUartConfig.UartId), UART_TXFIFO_EMPTY_INT_ENA);
+#ifndef UART_TX_BRK_DONE_INT_ENA
+#   define UART_TX_BRK_DONE_INT_ENA 0
+#endif // ndef | UART_TX_BRK_DONE_INT_ENA
+
+#ifndef UART_INTR_MASK
+#   if defined(ARDUINO_ARCH_ESP32)
+#       define UART_INTR_MASK uint32_t((1 << 18) - 1)
+#   elif defined(ARDUINO_ARCH_ESP8266)
+#       define UART_INTR_MASK uint32_t((1 << 8) - 1)
+#   endif //  defined(ARDUINO_ARCH_ESP8266)
+#endif // UART_INTR_MASK
+
+#if defined(ARDUINO_ARCH_ESP32)
+#   define WeNeedAtimer false
+#else
+#define WeNeedAtimer (OutputUartConfig.NumBreakBitsAfterIntensityData || OutputUartConfig.NumExtendedStartBits)
+#endif // defined(ARDUINO_ARCH_ESP32)
+
+#define ClearUartInterrupts   WRITE_PERI_REG(UART_INT_CLR(OutputUartConfig.UartId), UART_INTR_MASK);
```

Comment:
Here's another example where this psuedo-function macro (which presumes that there is a local variable of a given name) should likely become an `gnu_inline` function.

The GCC attributes of interest include:
* `gnu_inline` -- "In no case is the function compiled as a standalone function, not even if you take its address explicitly" per the [docs](https://gcc.gnu.org/onlinedocs/gcc-8.4.0/gcc/Common-Function-Attributes.html#Common-Function-Attributes).
* `const` (more strict than even `pure` attribute)

For example, it might look something close to the following:
```C++
inline void ClearUartInterrupts(OutputUartConfig_t& config) __attribute__(( gnu_inline, const )) {
    WRITE_PERI_REG(UART_INT_CLR(config.UartId), UART_INTR_MASK);
}
inline void DisableUartInterrupts(OutputUartConfig_t& config) __attribute__(( gnu_inline, const )) {
    CLEAR_PERI_REG_MASK(UART_INT_ENA(config.UartId), UART_INTR_MASK);
}
```

Reply:
Those functions need to be in IRAM which the macro nicely solved. At any rate, I converted them to IRAM inline functions. However, this is pretty much a waste of time and does not enhance safety or maintainability.

Reply:
I recognize you as an expert coder and your strong architectural skills.

I did not intend to offend; I have tried to only comment on items where maintainability or clarity would be increased, and worked hard to meet your apparent goals in those comments.

For example, I'm confused by the leading statement about IRAM... If using `gnu_inline` and `const` attributes as I suggested, then `IRAM_ATTR` wouldn't apply, because the function is never instantiated (can't even get a function pointer to it) ... it's forced to act substantially identically to a function-like macro.  If those attributes are not applied, then of course the `IRAM_ATTR` would be needed.  I noticed you chose not to use the attributes.
**_Can you help me understand the choice not to use them, as they would have provided the behavior you seemed to be looking for?_**

Can you help me understand your statement, "does not enhance safety or maintainability".
* Inline function has type safety for its parameters, Macro has no such type safety
* The macro was only intended for use within the class / in combination with the class object
* The macro was in **_common_** header file, and presumed a **_specific local variable name_**
Wouldn't these attributes show greater safety and maintainability from an inline function?

Reply:
- Inline function has type safety for its parameters, Macro has no such type safety
Actually, other than the enqueue macro, the macros do not have parameters. They count on well known data elements within a single class. 
- The macro was only intended for use within the class / in combination with the class object
Correct. They are very focused and therefore forced to stay very simple.
- The macro was in common header file, and presumed a specific local variable name
Correct. They are very focused and therefore forced to stay very simple.
- Wouldn't these attributes show greater safety and maintainability from an inline function?
Since the functionality in a macro is kept VERY simple, converting to inline actually provides very little benefit. Where more complex actions and potentially decisions are made, I agree a function is the better choice. Where an often repeated one line operation is performed where no decisions are made and no response is expected, Where no parameters are passed in, a macro is fine.

RE: IRAM_ATTR on inline. I was getting compiler issues and did not want to spend the time to resolve them. I will look at this again when I have more time.

Reply:
I think we are much closer aligned than it may have seemed.  

Since this was intended only for use within that one class, would you have considered moving the macro to the implementation (.cpp) file, rather than the header (.h/.hpp) file?

If so, although I would have written it as an inline function, I likely would not have mentioned it, as its scope would be limited to that single file.  (_Just another alternative, to support the preference for macros._)

P.S. - the functions may need to be only defined in the `.cpp` file, as otherwise the registers are not defined.  Here's an example that compiles, when placed in the implementation file:

```C++
static inline __attribute__(( always_inline, gnu_inline ))
void IRAM_ATTR EnableUartInterrupts(c_OutputUart::OutputUartConfig_t& config) {
    SET_PERI_REG_MASK  (UART_INT_ENA(config.UartId), UART_TXFIFO_EMPTY_INT_ENA);
}
static inline __attribute__(( always_inline, gnu_inline ))
void IRAM_ATTR DisableUartInterrupts(c_OutputUart::OutputUartConfig_t& config)  {
    CLEAR_PERI_REG_MASK(UART_INT_ENA(config.UartId), UART_TXFIFO_EMPTY_INT_ENA);
}
static inline __attribute__(( always_inline, gnu_inline ))
void IRAM_ATTR ClearUartInterrupts(c_OutputUart::OutputUartConfig_t& config)  {
    WRITE_PERI_REG     (UART_INT_CLR(config.UartId), UART_INTR_MASK);
}
```

Reply:
If they are coded as members of the uart class then I am ok with it. I am not ok with passing in those parameters.